### PR TITLE
Cupertino pull to refresh part 1: sliver and a simple indicator widget builder

### PR DESCRIPTION
**Pull request by xster**
#11661 

Next part: create the iOS 11 large title nav bar style refresh control inside the nav bar

### CODE REVIEW

File: packages/flutter/.vscode/tasks.json [hunk 1]
```diff
@@ -1,16 +1,20 @@
 {
-  "version": "0.1.0",
+  "version": "2.0.0",
```
Comment (xster) [inline packages/flutter/.vscode/tasks.json]: Drive-by VSCode helper update that makes test run in the editor

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+    Widget child,
+  }) : super(child: child);
+
+  /// The amount of space the indicator should occupy in the sliver in a
+  /// resting state when in the refreshing mode.
+  final double refreshIndicatorLayoutExtent;
+  /// _RenderCupertinoRefreshSliver will paint the child in the available
+  /// space either way but this instructs the _RenderCupertinoRefreshSliver
+  /// on whether to also occupy any layoutExtent space or not.
+  final bool hasLayoutExtent;
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: these should use `//` not `///` since they're not "camera ready" (see style guide for details)

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+      refreshIndicatorExtent: refreshIndicatorLayoutExtent,
+      hasLayoutExtent: hasLayoutExtent,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, covariant _RenderCupertinoRefreshSliver renderObject) {
+    renderObject
+        ..refreshIndicatorLayoutExtent = refreshIndicatorLayoutExtent
+        ..hasLayoutExtent = hasLayoutExtent;
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: indent by two fewer spaces for consistency

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+  /// not.
+  bool get hasLayoutExtent => _hasLayoutExtent;
+  bool _hasLayoutExtent;
+  set hasLayoutExtent(bool value) {
+    if (value == _hasLayoutExtent)
+      return;
+    _hasLayoutExtent = value;
+    markNeedsLayout();
+  }
+
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: if these properties can't be null, assert non-null in the constructor and the setters

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+  void applyPaintTransform(RenderObject child, Matrix4 transform) {}
+}
+
+/// The current state of the refresh control.
+///
+/// Passed into the [RefreshControlIndicatorBuilder] builder function so
+/// users can show different UI in different modes.
+enum RefreshIndicatorMode {
+  /// Initial state, when not being overscrolled into. Or after the overscroll
+  /// is canceled or after done and the sliver retracted away.
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: "into. Or" -> "into, or"

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+  refresh,
+  /// While the indicator is animating away after refreshing.
+  done,
+}
+
+/// A builder function that can create a different widget to show in the refresh
+/// indicator space depending on the current state of the refresh control and
+/// the space available.
+///
+/// The `refreshTriggerPullDistance`, `refreshIndicatorExtent` parameters are
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: "," -> "and"

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+  /// Dragged far enough that the onRefresh callback will run and the dragged
+  /// displacement is not yet at the final refresh resting state.
+  armed,
+  /// While the onRefresh task is running.
+  refresh,
+  /// While the indicator is animating away after refreshing.
+  done,
+}
+
+/// A builder function that can create a different widget to show in the refresh
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: Signature for a builder...

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+    this.refreshTriggerPullDistance: _kDefaultRefreshTriggerPullDistance,
+    this.refreshIndicatorExtent: _kDefaultRefreshIndicatorExtent,
+    this.builder: buildSimpleRefreshIndicator,
+    this.onRefresh,
+  }) : assert(refreshTriggerPullDistance != null && refreshTriggerPullDistance > 0),
+       assert(refreshIndicatorExtent != null && refreshIndicatorExtent >= 0),
+       assert(
+         refreshTriggerPullDistance >= refreshIndicatorExtent,
+         'The refresh indicator cannot take more space in its final state '
+         'than the amount initially created by overscrolling'
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: end messages with a period

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+  /// state will be drawn before going immediately to the [RefreshIndicatorMode.done]
+  /// where the sliver will start retracting.
+  final RefreshCallback onRefresh;
+
+  static const double _kDefaultRefreshTriggerPullDistance = 100.0;
+  static const double _kDefaultRefreshIndicatorExtent = 60.0;
+
+  /// Retrieve the current state of the CupertinoRefreshControl. The same as the
+  /// state that gets passed into the [builder] function. Used for testing.
+  @visibleForTesting
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: prefer @protected

Comment (Hixie) [reply packages/flutter/lib/src/cupertino/refresh.dart]: i'd rather we didn't expose this. You can get the state in the test by getting the widget by string (use a predicate finder that examines the runtimeType.toString) then getting its state as dynamic (tester.state).

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 10]
```diff
@@ -0,0 +1,514 @@
...
+                  min(pulledExtent / refreshIndicatorExtent, 1.0)
+                ),
+                child: const CupertinoActivityIndicator(radius: 14.0),
+              ),
+      ),
+    );
+  }
+
+  @override
+  _CupertinoRefreshControlState createState() {
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/refresh.dart]: this should eba one liner for consistency

**Comment by Hixie** [comment]
Overall approach looks good. I'll let someone else on the team do a more detailed review.

File: packages/flutter/lib/src/cupertino/activity_indicator.dart [hunk 11]
```diff
@@ -18,20 +20,25 @@ class CupertinoActivityIndicator extends StatefulWidget {
   const CupertinoActivityIndicator({
     Key key,
     this.animating: true,
+    this.radius: _kDefaultIndicatorRadius,
   }) : assert(animating != null),
+       assert(radius != null && radius > 0),
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/activity_indicator.dart]: minor thing: generally speaking, it's better to have two separate asserts than an assert with two expressions &&'ed together, because then when the assert fails, you can more immediately tell what exactly the problem is.
(Applies to various asserts in this PR.)

**Comment by Hixie** [comment]
cc @gspencergoog or @goderbauer for review

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_refresh_demo.dart [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+List<List<String>> contacts = <List<String>>[
```
Comment (gspencergoog) [inline examples/flutter_gallery/lib/demo/cupertino/cupertino_refresh_demo.dart]: A list of U.S. presidents is a little U.S.-centric.  Maybe a list of inventors or Nobel prize winners or something more international?

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+typedef Future<void> RefreshCallback();
+
+/// A sliver widget implementing the iOS-style pull to refresh content control.
+///
+/// When inserted as the first sliver in a scroll view or behind other slivers
+/// that still lets the scrollable overscroll in front of this sliver (such as
+/// the [CupertinoSliverNavigationBar], this widget will:
+///
+///  * Let the user draw inside the overscrolled area via the passed in [builder].
+///  * Trigger the provide [onRefresh] function when overscroll far enough to pass
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: provide --> provided
overscroll --> overscrolled

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+///    [refreshTriggerPullDistance].
+///  * Continue to hold [refreshIndicatorExtent] amount of space for the [builder]
+///    to keep drawing inside as the [Future] returned by [onRefresh] processes.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: inside --> inside of,

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+///  * Scroll away once the [onRefresh] [Future] completes.
+///
+/// The [builder] function will be informed of the current [RefreshIndicatorMode]
+/// when invoking it except in the [RefreshIndicatorMode.inactive] state when
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: it except --> it, except

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+/// no space is available and nothing needs to be built. The [builder] function
+/// will otherwise be continuously invoked as the amount of space available
+/// changes from overscroll, as the sliver scrolls away after the [onRefresh]
+/// task is done etc.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: done etc. --> done, etc.

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+///
+/// Only one refresh can be triggered at once until the previous refresh has
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: triggered at once until --> triggered until

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 18]
```diff
@@ -0,0 +1,520 @@
...
+  /// Must not be null, must be larger than 0 and larger than [refreshIndicatorExtent].
+  ///
+  /// When overscrolled past this distance, [onRefresh] will be called if not
+  /// null and the [builder] will build in the [RefreshIndicatorMode.armed] state.
+  final double refreshTriggerPullDistance;
+
+  /// The amount of space the refresh indicator sliver will keep holding while
+  /// [onRefresh]'s [Future] is still running.
+  ///
+  /// Must not be null and must be positive. But can be 0, in which case the
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: positive. But --> positive, but

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 19]
```diff
@@ -0,0 +1,520 @@
...
+  /// sliver will start retracting back to 0 as soon as the refresh started.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: started --> is started

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 20]
```diff
@@ -0,0 +1,520 @@
...
+  ///
+  /// Must be smaller than [refreshTriggerPullDistance] since the sliver shouldn't
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: [refreshTriggerPullDistance] since --> [refreshTriggerPullDistance], since

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 21]
```diff
@@ -0,0 +1,520 @@
...
+  /// grow more after triggering the refresh.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: grow more after --> grow after

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 22]
```diff
@@ -0,0 +1,520 @@
...
+  final double refreshIndicatorExtent;
+
+  /// A builder that's called as this sliver's size changes and as the state
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: changes and --> changes, and

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 23]
```diff
@@ -0,0 +1,520 @@
...
+  /// changes.
+  ///
+  /// A default simple Twitter-style pull-to-refresh indicator is provided if
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: I'm not sure the "Twitter-style" adds anything (I, for one, have never seen Twitter's indicator, so I have no idea what it looks like).  I think just saying "A default pull-to-refresh indicator..." is fine, and doesn't endorse a product.

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 24]
```diff
@@ -0,0 +1,520 @@
...
+  /// not specified.
+  ///
+  /// Can be set to null, in which case nothing will be drawn in the overscrolled
+  /// space.
+  ///
+  /// Will not be called when the available space is 0 such as before any
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: 0 such --> zero, such

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 25]
```diff
@@ -0,0 +1,520 @@
...
+  /// overscroll.
+  final RefreshControlIndicatorBuilder builder;
+
+  /// Callback invoked when pulled by [refreshTriggerPullDistance].
+  ///
+  /// If provided, must return a [Future] which will keep the indicator in the
+  /// [RefreshIndicatorMode.refresh] state until the [Future] completes.
+  ///
+  /// Can be null in which case a single frame of [RefreshIndicatorMode.armed]
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: null in --> null, in

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 26]
```diff
@@ -0,0 +1,520 @@
...
+  /// state will be drawn before going immediately to the [RefreshIndicatorMode.done]
+  /// where the sliver will start retracting.
+  final RefreshCallback onRefresh;
+
+  static const double _kDefaultRefreshTriggerPullDistance = 100.0;
+  static const double _kDefaultRefreshIndicatorExtent = 60.0;
+
+  /// Retrieve the current state of the CupertinoRefreshControl. The same as the
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: CupertinoRefreshControl. The same as the --> CupertinoRefreshControl, which is the same as the

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 27]
```diff
@@ -0,0 +1,520 @@
...
+  RefreshIndicatorMode refreshState;
+  // [Future] returned by the widget's `onRefresh`.
+  Future<void> refreshTask;
+  // The amount of space available from the inner indicator box's perspective.
+  //
+  // The value is the sum of the sliver's layout extent and the overscroll
+  // (which partially gets transfered into the layout extent when the refresh
+  // triggers).
+  //
+  // The value is independent from the slivers scrollOffset. i.e. it's still
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: scrollOffset. i.e. it's --> scrollOffset: it's

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 28]
```diff
@@ -0,0 +1,520 @@
...
+      case RefreshIndicatorMode.drag:
+        if (lastIndicatorExtent == 0) {
+          return RefreshIndicatorMode.inactive;
+        } else if (lastIndicatorExtent < widget.refreshTriggerPullDistance) {
+          return RefreshIndicatorMode.drag;
+        } else {
+          if (widget.onRefresh != null) {
+            HapticFeedback.mediumImpact();
+            // Call onRefresh after this frame finished since the function is
+            // user supplied and we're always here in the middle of a the sliver's
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: of a the --> of the
(or 'of a', depending on what you meant)

File: packages/flutter/lib/src/cupertino/refresh.dart [hunk 29]
```diff
@@ -0,0 +1,520 @@
...
+        // can feel sluggish if not going all the way back to 0.0 prevented
+        // a subsequent pull-to-refresh from starting.
+        if (lastIndicatorExtent >
+            widget.refreshTriggerPullDistance * _kInactiveResetOverscrollFraction) {
+          return RefreshIndicatorMode.done;
+        } else {
+          nextState = RefreshIndicatorMode.inactive;
+        }
+        break;
+      default:
```
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/refresh.dart]: Don't put a default here.  You've handled all the cases, so it shouldn't be necessary, and will highlight a missing case if more modes are added in the future.

File: packages/flutter/test/cupertino/refresh_test.dart [hunk 30]
```diff
@@ -0,0 +1,1260 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/test/cupertino/refresh_test.dart]: 2018?

File: packages/flutter/test/cupertino/refresh_test.dart [hunk 31]
```diff
@@ -0,0 +1,1260 @@
...
+        expect(
+          tester.getTopLeft(find.widgetWithText(Center, '-1')).dy,
+          moreOrLessEquals(-175.38461538461536),
+        );
+        expect(
+          tester.getBottomLeft(find.widgetWithText(Center, '-1')).dy,
+          moreOrLessEquals(-115.38461538461536),
+        );
+
+        // Complete the task while scrolled await.
```
Comment (gspencergoog) [inline packages/flutter/test/cupertino/refresh_test.dart]: scrolled await. --> scrolled.
(unless you meant something else)

File: packages/flutter/.vscode/tasks.json (hunk 1 above)
Comment (xster) [reply packages/flutter/.vscode/tasks.json]: Removed this file since we said we don't want vscode stuff

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 2 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: Ah ok, review and replaced all the comments in the file

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 3 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: Ummm I just realized https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo doesn't say to use 4 spaces indents anywhere. Wonder if it changed recently. Done.

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 4 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: Done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 5 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 6 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 7 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 8 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 9 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: ya, I was trying to figure out how to not expose this too. I can get to the state but the issue is getting the field in the state since the state class is private

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 10 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/activity_indicator.dart (hunk 11 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart]: done

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_refresh_demo.dart (hunk 12 above)
Comment (xster) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_refresh_demo.dart]: ah ya, that's a good idea. I'll try to find another list in a separate PR. Just need to find something with specific dates so it looks similar to the native contacts app.

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 13 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: that sentence made no sense :S thanks

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 14 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 15 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 16 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 17 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 18 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 19 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 20 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 21 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 22 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 23 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: It's not about twitter per se. The pull to refresh UI pattern is upstreamed into the iOS framework from the twitter app originally. It's just referring to the 'original' style pull to refresh as opposed to the iOS 11 new style which will be added in the next pr

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 24 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: Using digit 0 in the rest of the doc. Keeping for consistency

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 25 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 26 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: done

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 28 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: ah thanks

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 29 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: good idea, done

File: packages/flutter/test/cupertino/refresh_test.dart (hunk 31 above)
Comment (xster) [reply packages/flutter/test/cupertino/refresh_test.dart]: doh

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 27 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: I meant more 'in other words'. ':' sounds like the rest is referring to scrollOffset rather than lastIndicatorExtent

File: packages/flutter/lib/src/cupertino/activity_indicator.dart (hunk 11 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/cupertino/activity_indicator.dart]: Hmm.  Looks like you missed this one..

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 24 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/cupertino/refresh.dart]: Well, OK, but most writing style guides agree that numerals under 10 should be written out, rather than using digits, and since this is user-facing text, grammar and style is important.

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 27 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/cupertino/refresh.dart]: Then how about:
The value of `lastIndicatorExtent` doesn't change when the sliver scrolls away without retracting; it is independent from the sliver's `scrollOffset`.

When I read yours, it was unclear what "it's" was referring to.

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d
(modulo nits)

File: examples/flutter_gallery/lib/demo/cupertino/cupertino_refresh_demo.dart (hunk 12 above)
Comment (Hixie) [reply examples/flutter_gallery/lib/demo/cupertino/cupertino_refresh_demo.dart]: Or just pick a list of presidents or something from another country. :-)

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 27 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: sg

File: packages/flutter/lib/src/cupertino/refresh.dart (hunk 24 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/refresh.dart]: Seems like there are 98 instances of `0` in our dartdocs and 103 `zero` in our dartdocs :S

I changed the `0` in this file to `0.0` when I meant the double value exactly and `zero` when I meant the zero value conceptually like here.

**Comment by realcarlos** [comment]
Thank you , it will be great if there is a pull-up-load-more demo .

**Comment by SemonCat** [comment]
any function can trigger refresh programming way ?
